frontend/appflowy_flutter/integration_test/desktop/grid/grid_create_row_test.dart

```diff
@@ -0,0 +1,182 @@
+import 'package:appflowy/plugins/database/grid/presentation/widgets/filter/choicechip/checkbox.dart';
+import 'package:appflowy_backend/protobuf/flowy-database2/protobuf.dart';
+import 'package:appflowy_backend/protobuf/flowy-folder/view.pbenum.dart';
+import 'package:collection/collection.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:integration_test/integration_test.dart';
+
+import '../../shared/database_test_op.dart';
+import '../../shared/util.dart';
+import 'grid_test_extensions.dart';
+
+void main() {
+  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
+
+  group('grid create row test:', () {
+    testWidgets('from the bottom', (tester) async {
+      await tester.initializeAppFlowy();
+      await tester.tapAnonymousSignInButton();
+
+      await tester.createNewPageWithNameUnderParent(layout: ViewLayoutPB.Grid);
+
+      final expected = tester.getGridRows();
+
+      // create row
+      await tester.tapCreateRowButtonInGrid();
+
+      final actual = tester.getGridRows();
+      expect(actual.slice(0, 3), orderedEquals(expected));
+      expect(actual.length, equals(4));
+      tester.assertNumberOfRowsInGridPage(4);
+    });
+
+    testWidgets('from a row\'s menu', (tester) async {
+      await tester.initializeAppFlowy();
+      await tester.tapAnonymousSignInButton();
+
+      await tester.createNewPageWithNameUnderParent(layout: ViewLayoutPB.Grid);
+
+      final expected = tester.getGridRows();
+
+      // create row
+      await tester.hoverOnFirstRowOfGrid();
+      await tester.tapCreateRowButtonAfterHoveringOnGridRow();
+
+      final actual = tester.getGridRows();
+      expect([actual[0], actual[2], actual[3]], orderedEquals(expected));
+      expect(actual.length, equals(4));
+      tester.assertNumberOfRowsInGridPage(4);
+    });
+
+    testWidgets('with sort configured', (tester) async {
+      await tester.openTestDatabase(v069GridFileName);
+
+      // get grid data
+      final unsorted = tester.getGridRows();
+
+      // add a sort
+      await tester.tapDatabaseSortButton();
+      await tester.tapCreateSortByFieldType(FieldType.RichText, 'Name');
+
+      final sorted = [
+        unsorted[7],
+        unsorted[8],
+        unsorted[1],
+        unsorted[9],
+        unsorted[11],
+        unsorted[10],
+        unsorted[6],
+        unsorted[12],
+        unsorted[2],
+        unsorted[0],
+        unsorted[3],
+        unsorted[5],
+        unsorted[4],
+      ];
+
+      final actual = tester.getGridRows();
+      expect(actual, orderedEquals(sorted));
+
+      // create row
+      await tester.tapCreateRowButtonInGrid();
+
+      // TODO(RS): handle create row with sort active before writing test logic
+    });
+
+    testWidgets('with filter configured', (tester) async {
+      await tester.openTestDatabase(v069GridFileName);
+
+      // get grid data
+      final original = tester.getGridRows();
+
+      // create a filter
+      await tester.tapDatabaseFilterButton();
```

Comment:
Failure caused by incorrect refresh of the frontend, leading to the new row being inserted at the wrong index. Refreshing the database view shows the correct row order.

Reply:
Resolved